### ISSUE TITLE
"str" cannot be assigned to "AnyStr"

### ISSUE THREAD

**Issue by bschnurr** [microsoft/pyright#1834]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
in strict mode
![image](https://user-images.githubusercontent.com/1946977/117377040-33ee3800-ae87-11eb-8afb-9a0721a02dfc.png)
**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
def foo(url: AnyStr, seed: str):
    url = seed
    print(url)
    pass
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

**Additional context**

**Comment by jakebailey** [comment]
This is how the `AnyStr` typevar works. Consider:

```python
def identity(x: AnyStr) -> AnyStr:
    return x

print(type(identity("foo")))   # str
print(type(identity(b"foo")))  # bytes
```

But if the function were changed to return a str:

```
def identity(x: AnyStr) -> AnyStr:
    return "something"

print(type(identity("foo")))   # str
print(type(identity(b"foo")))  # str
```

**Comment by erictraut** [comment]
The behavior is correct in this case. Perhaps you meant to annotate `url` with `Union[str, bytes]` instead? The assignment statement would then be legal.

**Comment by jakebailey** [comment]
I think the case he's looking at right now is this: https://github.com/iotaledger/iota.py/blob/f596c1ac0d9bcbceda1cf6109cd921943a6599b3/iota/types.py#L27

An `AnyStr` in a Union that causes troubles later.

**Comment by erictraut** [comment]
Yeah, that should probably be changed to `TrytesCompatible = Union[str, bytes, bytearray, 'TryteString']`. The use of a constrained TypeVar in this context doesn't make sense.

**Comment by jakebailey** [comment]
Any idea why there's no error for this? I can't imagine a legal use of a type var like this... There's no scope binding it.

**Comment by erictraut** [comment]
`TrytesCompatible = Union[AnyStr, bytearray, 'TryteString']` defines a generic type alias, which can be legitimate. To use it properly, you'd need to provide a type argument that is compatible with the `AnyStr` type variable.

`TrytesCompatible[str]` or `TrytesCompatible[bytes]` or `TrytesCompatible[AnyStr]` would all be allowed.

But I don't think that's what the author of this code had in mind.

**Comment by jakebailey** [comment]
Yeah, makes sense. I assume they thought they had found a shorter way to write "str or bytes" 🙂.

**Comment by erictraut** [comment]
Yeah. Constrained TypeVars are very misunderstood. They're misused almost as often as they're used incorrectly. I think it's unfortunate that they're part of the language.

**Comment by bschnurr** [comment]
> "str or bytes

yup that's what I think when I see AnyStr

**Comment by bschnurr** [comment]
That fixes a bunch of things thanks!
